overlord/snapstate/handlers.go

```diff
@@ -416,6 +425,14 @@ func (m *SnapManager) doDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 
 	meter := NewTaskProgressAdapterUnlocked(t)
 	targetFn := snapsup.MountFile()
+
+	ctx := tomb.Context(nil)
+	st.Lock()
+	rate := rateLimited(st)
+	st.Unlock()
+	if rate > 0 {
+		ctx = context.WithValue(ctx, "rate-limit", float64(rate))
```

Comment:
The context trap. IIRC it's better to not use basic types as context keys, so the context magic should look more like this:

```
type RateLimitKey int
const rateLimitKey RateLimitKey = 1

func WithRateLimit(ctx context.Context, rate float64) context.Context {
    return context.WithValue(ctx, rateLimitKey, rate)
}

func GetRateLimit(ctx context.Context) float64 {
    if val, ok := ctx.Value(rateLimitKey).(float64); ok {
        return val
    }
    return -1
}
```

https://play.golang.org/p/JQLs3r5r_Cy

Reply:
Totally agreed. I was just lazy for this prototype but yeah, this needs to be a type just like you describe it here!